pkg/util/ipfilter/layer4ipfilters.go

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ipfilter
+
+import (
+	"reflect"
+	"sync/atomic"
+)
+
+type (
+	// Layer4IpFilters layer4 ip filters
+	Layer4IpFilters struct {
+		rules atomic.Value
+	}
+
+	ipFiltersRules struct {
+		spec     *Spec
+		ipFilter *IPFilter
+	}
+)
+
+// NewLayer4IPFilters create layer4 ip filters
+func NewLayer4IPFilters(spec *Spec) *Layer4IpFilters {
+	if spec == nil {
+		return &Layer4IpFilters{}
+	}
+
+	m := &Layer4IpFilters{}
+	m.rules.Store(&ipFiltersRules{
+		spec:     spec,
+		ipFilter: New(spec),
+	})
+	return m
+}
+
+// AllowIP check whether the IP is allowed to pass
+func (i *Layer4IpFilters) AllowIP(ip string) bool {
+	rules := i.rules.Load().(*ipFiltersRules)
```

Comment:
UDP Server seems to expect that `i.rules` is not empty. When I create `UDPServer` without **ipFilters** specified:
```yaml
kind: UDPServer
name: udp-server-demo
port: 10070
maxConns: 1024
hasResponse: true
clientIdleTimeout: 30000
serverIdleTimeout: 30000
pool:
  servers:
    - url: '127.0.0.1:9095'
  loadBalance:
    policy: roundRobin
```
and then open UDP connection, I got this error:

```
panic: interface conversion: interface {} is nil, not *ipfilter.ipFiltersRules

goroutine 5880 [running]:
github.com/megaease/easegress/pkg/util/ipfilter.(*Layer4IpFilters).AllowIP(0xc000b272cc, {0xc00080f7d5, 0xc0014b0000})
	github.com/megaease/easegress/pkg/util/ipfilter/layer4ipfilters.go:53 +0xa5
github.com/megaease/easegress/pkg/object/udpproxy.(*runtime).startServer.func1()
	github.com/megaease/easegress/pkg/object/udpproxy/runtime.go:118 +0x12c
created by github.com/megaease/easegress/pkg/object/udpproxy.(*runtime).startServer
	github.com/megaease/easegress/pkg/object/udpproxy/runtime.go:92 +0x245
```

This does not happen with TCPServer, only in the case of UDPServer.

Reply:
It's my mistake.
I will fix it.

Reply:
I tried the fix in https://github.com/megaease/easegress/pull/234/commits/af9a795ef79cae3608f635743fca527c97153e4a and it seems to help with `UDPProxy` not having ipFilters. But now the error comes when creating `TCPProxy` without **ipFilters**:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1b3bdea]

goroutine 327 [running]:
github.com/megaease/easegress/pkg/util/ipfilter.(*Layer4IpFilters).ReloadRules(0x0, 0x0)
	github.com/megaease/easegress/pkg/util/ipfilter/layer4ipfilters.go:63 +0x12a
github.com/megaease/easegress/pkg/object/tcpproxy.(*runtime).reload(0xc002a909c0, 0xc0007b0680)
	github.com/megaease/easegress/pkg/object/tcpproxy/runtime.go:128 +0x68
github.com/megaease/easegress/pkg/object/tcpproxy.(*runtime).handleEventReload(...)
	github.com/megaease/easegress/pkg/object/tcpproxy/runtime.go:261
github.com/megaease/easegress/pkg/object/tcpproxy.(*runtime).fsm(0xc002a909c0)
	github.com/megaease/easegress/pkg/object/tcpproxy/runtime.go:112 +0x9c
created by github.com/megaease/easegress/pkg/object/tcpproxy.newRuntime
	github.com/megaease/easegress/pkg/object/tcpproxy/runtime.go:91 +0x187
```
I think it's better to always create the `Layer4IpFilters` AND `ipFiltersRules` structs, like this

```go
func NewLayer4IPFilters(spec *Spec) *Layer4IpFilters {
       m := &Layer4IpFilters{}
       rules := &ipFiltersRules{}
       if spec != nil {
               rules = &ipFiltersRules{
                       spec:     spec,
                       ipFilter: New(spec),
               }
        }
        m.rules.Store(rules)
        return m
 }
```
Now it should avoid the both nil pointer exceptions and work as expected... you can verify if that's the case.

Reply:
So sad for my modification.
I will take your suggestions and complete the repair as soon as possible.